File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.py
class Solution:
    def takeCharacters(self,s,k):
        d={c:s.count(c)-k for c in'abc'}
        if min(d.values())<0:return-1
        h,m,l={c:0 for c in'abc'},0,0
        for r,c in enumerate(s):
            h[c]+=1
            while h[c]>d[c]:h[s[l]]-=1;l+=1
            m=max(m,r-l+1)
        return len(s)-m

File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.cpp
class Solution {
public:
    int takeCharacters(string s, int k) {
        vector<int> d(3), h(3);
        for (char c : s) d[c-'a']++;
        for (int i = 0; i < 3; i++) if ((d[i] -= k) < 0) return -1;
        
        int m = 0, l = 0;
        for (int r = 0; r < s.length(); r++) {
            h[s[r]-'a']++;
            while (l <= r && h[s[r]-'a'] > d[s[r]-'a']) h[s[l++]-'a']--;
            m = max(m, r - l + 1);
        }
        return s.length() - m;
    }
};